radix-engine/src/blueprints/resource/non_fungible/non_fungible_vault.rs

```diff
@@ -0,0 +1,663 @@
+use crate::blueprints::resource::*;
+use crate::errors::ApplicationError;
+use crate::errors::RuntimeError;
+use crate::kernel::heap::{DroppedBucket, DroppedBucketResource};
+use crate::kernel::kernel_api::{KernelNodeApi, KernelSubstateApi};
+use crate::types::*;
+use native_sdk::runtime::Runtime;
+use radix_engine_interface::api::types::*;
+use radix_engine_interface::api::{ClientApi, ClientSubstateApi, LockFlags};
+use radix_engine_interface::blueprints::resource::*;
+
+#[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor)]
+pub struct NonFungibleVaultInfoSubstate {
+    pub resource_address: ResourceAddress,
+    pub id_type: NonFungibleIdType,
+}
+
+pub struct NonFungibleVaultBlueprint;
+
+impl NonFungibleVaultBlueprint {
+    fn check_amount(amount: &Decimal) -> bool {
+        !amount.is_negative() && amount.0 % BnumI256::from(10i128.pow(18)) == BnumI256::from(0)
+    }
+
+    fn get_info<Y>(
+        receiver: &RENodeId,
+        api: &mut Y,
+    ) -> Result<NonFungibleVaultInfoSubstate, RuntimeError>
+    where
+        Y: KernelNodeApi + KernelSubstateApi + ClientApi<RuntimeError>,
+    {
+        let handle = api.sys_lock_substate(
+            receiver.clone(),
+            SubstateOffset::Vault(VaultOffset::Info),
+            LockFlags::read_only(),
+        )?;
+        let info: &NonFungibleVaultInfoSubstate = api.kernel_get_substate_ref(handle)?;
+        let info = info.clone();
+        api.sys_drop_lock(handle)?;
+        Ok(info)
+    }
+
+    pub fn take<Y>(
+        receiver: &RENodeId,
+        amount: &Decimal,
+        api: &mut Y,
+    ) -> Result<Bucket, RuntimeError>
+    where
+        Y: KernelNodeApi + KernelSubstateApi + ClientApi<RuntimeError>,
+    {
+        // Check amount
+        let info = Self::get_info(receiver, api)?;
+        if !Self::check_amount(amount) {
+            return Err(RuntimeError::ApplicationError(
+                ApplicationError::VaultError(VaultError::InvalidAmount),
+            ));
+        }
+
+        // Take
+        let taken = NonFungibleVault::take(receiver, *amount, api)?;
+
+        // Create node
+        let bucket_id = api.new_object(
```

Comment:
Ditto - all this code which creates buckets feels like it should probably live as a bucket constructor.